system_metrics_collector/src/topic_statistics_collector/subscriber_topic_statistics.hpp

```diff
@@ -0,0 +1,175 @@
+
+  virtual ~SubscriberTopicStatistics() = default;
+
+  /**
+   * Returns a pretty printed status representation of this class
+   *
+   * @return a string detailing the current status
+   */
+  std::string GetStatusString() const override;
+
+  /**
+   * Starts the node.
+   *
+   * @param state input state unused
+   * @return CallbackReturn success if start returns true, error otherwise
+   */
+  rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn on_activate(
+    const rclcpp_lifecycle::State & state);
+
+  /**
+   * Stops the node.
+   *
+   * @param input state unused
+   * @return CallbackReturn success if start returns true, error otherwise
+   */
+  rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn on_deactivate(
+    const rclcpp_lifecycle::State & state);
+
+  /**
+   * Stops the node and performs cleanup.
+   *
+   * @param input state unused
+   * @return CallbackReturn success
+   */
+  rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn on_shutdown(
+    const rclcpp_lifecycle::State & state);
+
+  /**
+   * Stops the node and attempts to perform cleanup.
+   *
+   * @param input state unused
+   * @return CallbackReturn success
+   */
+  rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn on_error(
+    const rclcpp_lifecycle::State & previous_state);
+
+  /**
+   * Creates ROS2 timers and a publisher for periodically triggering measurements
+   * and publishing MetricsMessages
+   *
+   * @return if setup was successful
+   */
+  bool SetupStart() override;
+
+  /**
+   * Stops the ROS2 timers that were created by SetupStart()
+   *
+   * @return if teardown was successful
+   */
+  bool SetupStop() override;
+
+  /**
+   * LifecyclePublisher publisher that is activated on SetupStart and deactivated on SetupStop().
+   */
+  rclcpp_lifecycle::LifecyclePublisher<metrics_statistics_msgs::msg::MetricsMessage>::SharedPtr
+    publisher_;
+
+private:
+  /**
+   * Publishes the statistics derived from the collected measurements (this is to be called via a
+   * ROS2 timer per the publish_period)
+   */
+  void PublishStatisticMessage() override;
+
+  /**
+   *  Callback function to execute when messages on subscribed topics are received
+   *
+   *  @param msg message received
+  **/
+  void CollectorCallback(const typename T::SharedPtr received_message);
+
+  /**
+   * Tracks the starting time of the statistics
+   */
+  rclcpp::Time window_start_;
+
+  /**
+   * The period used to publish measurement data
+   */
+  std::chrono::milliseconds publish_period_{0};
+
+  /**
+   * ROS2 timer used to publish measurement messages
+   */
+  rclcpp::TimerBase::SharedPtr publish_timer_;
+
+  /**
+  * Subscriber to listen to incoming messages on a topic
+  */
+  typename rclcpp::Subscription<T>::SharedPtr subscription_;
+
+  /**
+  * Subscription options to configure the subscriber
+  */
+  rclcpp::SubscriptionOptions subscription_options_;
```

Comment:
Could some of those attributes made const?

Reply:
Removed all unnecessary variables. The remaining are modified during object lifecycle so cannot be made const.